### DISCUSSION TITLE
Issues when attempting to resample 32Khz sample to 22.05khz

### DISCUSSION THREAD

**Discussion by Ewan-Stacey-BTO-IS** [psambit9791/jDSP#75] [Q&A] [answered]
I am trying to use the Resample class to resample audio data in a double array to 22.05Khz, the resampling code seems to work on 32khz files that are under two minutes long but seemingly fails on all other 32khz files but seems to work fine with 44.1kHz, 48kHz and 84kHz files of any size, here is the stack trace:
Exception stacktrace:
-647115
      com.github.psambit9791.jdsp.signal.Resample$_UpFIRDown.apply_filter(Resample.java:372)
      com.github.psambit9791.jdsp.signal.Resample$_UpFIRDown.access$100(Resample.java:312)
      com.github.psambit9791.jdsp.signal.Resample.resample_poly(Resample.java:282)
      com.github.psambit9791.jdsp.signal.Resample.resampleSignal(Resample.java:301)
      org.bto.acousticpipeline.services.WavFileImporter.importWavToFloatArray(WavFileImporter.java:138)

The code from my program:
```Java
if (originalFormat.getSampleRate() != desiredSampleRate) {
                Resample resampler = new Resample(desiredSampleRate.intValue(), (int) originalFormat.getSampleRate(), "constant");
                double[] preResamplingArray = convertFloatArrayToDouble(trimmedValues);
                ArrayList<double[]> splitResampledArrays = new ArrayList<>();
                int outerArraySize = (int) Math.ceil((double) preResamplingArray.length / (double) MAX_ARRAY_RESAMPLE_SIZE);
                for (int x = 0; x < outerArraySize; x++) {
                    int endOfRange;
                    if (x == (outerArraySize - 1)) {
                        endOfRange = preResamplingArray.length;
                    } else {
                        endOfRange = ((x + 1) * MAX_ARRAY_RESAMPLE_SIZE);
                    }
                    double[] splitArray = Arrays.copyOfRange(preResamplingArray, x * MAX_ARRAY_RESAMPLE_SIZE, endOfRange);
                    double[] resampledDoubleArray = resampler.resampleSignal(splitArray); //error occurs here
                    splitResampledArrays.add(resampledDoubleArray);
                }
                double[] recombinedDoubleArray = recombineDoubleArray(splitResampledArrays);
                resampledValues = convertDoubleArrayToFloat(recombinedDoubleArray);
            } else {
                resampledValues = trimmedValues;
            }
```
for reference:
`MAX_ARRAY_RESAMPLE_SIZE = 10000000` This is an arbitrary size, it's just that running whole files through the resampling process seems break the resampler so I break them into smaller arrays
`trimmedValues` is a `float` array that contains samples 

Relevant methods
```Java
public static double[] convertFloatArrayToDouble (float[] floatArray) {
        double[] doubleArray = new double[floatArray.length];
        
        int index = 0;
        
        for(float floatValue : floatArray) {
            doubleArray[index] = (double) floatValue;
            index++;
        }
        
        return doubleArray;
    }
```

The error occurs on `resampler.resampleSignal(splitArray);` so anything afterwards shouldn't effect it

We are specifically using this library because of the anti-aliasing unlike the default java resampling when using AudioInputStream

We know about the decimate function but since we are trying to convert to 22.05khz we do not think it will work for our purposes, additionally we sometime would need to up sample

I'm not really familiar with the mathematics of resampling, so I'm not really sure what could be going wrong

**Comment by psambit9791** [comment]
1. The usage of Resample is incorrect here. The utility reflects the method used in [scipy.resample_poly](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.resample_poly.html#:~:text=the%20FFT%20method.-,Notes,-This%20polyphase%20method). 
The first two parameters in the function invocation `Resample resampler = new Resample(desiredSampleRate.intValue(), (int) originalFormat.getSampleRate(), "constant")` should not be the target frequency and source frequency - instead they need to be the upscaling and downscaling ratio **as integers**. These can be computed using the GCD between the target and source frequencies.
Please refer to the hyperlink for how it works in *scipy* and please use the [documentation for Resample](https://javadoc.io/doc/com.github.psambit9791/jdsp/latest/com.github.psambit9791.jdsp/com/github/psambit9791/jdsp/signal/Resample.html) in case you want details on the parameters to be used.

2. Can you please share a sample file I can use to recreate the issue? This file needs to be the subsample (the smaller array from the original data) which I can read using the WAV class.

**Reply by Ewan-Stacey-BTO-IS** [reply]
Thank you very much for the response and explanation, I still seem to be having difficulties with 32000khz files with this method
I have changed the initialisation to be like this
```Java
int greatestCommonDivisor = originalFormat.getSampleRate() > desiredSampleRate ? greatestCommonDivisorEuclids(originalFormat.getSampleRate(), desiredSampleRate) 
                        : greatestCommonDivisorEuclids(desiredSampleRate, originalFormat.getSampleRate());

                System.out.println("GREATEST COMMON DIVISOR: " + greatestCommonDivisor);
                int ratioUpsample = (int) (desiredSampleRate / greatestCommonDivisor);
                System.out.println("RATIO UPSAMPLE: " + ratioUpsample);
                int ratioDownsample = (int) (originalFormat.getSampleRate() / greatestCommonDivisor);
                System.out.println("RATIO DOWNSAMPLE: " + ratioDownsample);
                Resample resampler = new Resample(ratioUpsample, ratioDownsample, "constant");
//The Sys outs are just to check the method was working as intended

//https://www.baeldung.com/java-greatest-common-divisor#bd-euclid's%20algorithm
    private static int greatestCommonDivisorEuclids (float greaterNumber, float smallerNumber) {
        if(smallerNumber == 0) {
            return (int) greaterNumber;
        }
        return greatestCommonDivisorEuclids(smallerNumber, greaterNumber % smallerNumber);
    }
```

The output for those system outs is
GREATEST COMMON DIVISOR: 50
RATIO UPSAMPLE: 441
RATIO DOWNSAMPLE: 640

The stack trace looks like this

Exception stacktrace:
-647115
      com.github.psambit9791.jdsp.signal.Resample$_UpFIRDown.apply_filter(Resample.java:372)
      com.github.psambit9791.jdsp.signal.Resample$_UpFIRDown.access$100(Resample.java:312)
      com.github.psambit9791.jdsp.signal.Resample.resample_poly(Resample.java:282)
      com.github.psambit9791.jdsp.signal.Resample.resampleSignal(Resample.java:301)
      org.bto.acousticpipeline.services.WavFileImporter.importWavToFloatArray(WavFileImporter.java:143)


For the program we are going to try to switch over to the non poly resample method though I can keep going with this one if it is of any help to you

As for the sound file, would just sending the WAV file as is to you be okay?

**Reply by psambit9791** [reply]
@Ewan-Stacey-BTO-IS 
Just a subsample of the WAV file would be good. Something which I can directly read without having the break it into smaller arrays.

**Reply by Ewan-Stacey-BTO-IS** [reply]
How would I go about getting a subsample, when I input the array into my code I only get a singular array of bytes which I manually convert to float data

**Reply by psambit9791** [reply]
Maybe just send me a small 3 seconds segment from the relevant WAV file. You can use any audio processing tool for this - maybe something like [veed.io](https://www.veed.io/tools/split-audio)?

**Reply by Ewan-Stacey-BTO-IS** [reply]
I can't upload WAV files to this thread, could I get an email address to send the file to please

**Reply by psambit9791** [reply]
Can you please use the 'Paste, drop, or click to add files' option?
Shown in this image.
![image](https://github.com/psambit9791/jdsp/assets/25612092/346df7cf-af15-466e-a308-7d58d9295344)

**Reply by Ewan-Stacey-BTO-IS** [reply]
I cannot as WAV files are not allowed to be uploaded by Github on the discussion threads
![image](https://github.com/psambit9791/jdsp/assets/156305373/4d5f9db3-dea4-4463-8ada-7f579903185d)
I could upload the file in a public repo and then send it to you that way I guess?

**Reply by psambit9791** [reply]
You can zip the file and upload it here itself. If that doesn't work, you can send it to me at sambitpaul1992@gmail.com

**Reply by Ewan-Stacey-BTO-IS** [reply]
Oh sorry 
[audio_file.zip](https://github.com/psambit9791/jdsp/files/13939111/audio_file.zip)
There we go

**Reply by Ewan-Stacey-BTO-IS** [reply]
Hey Sorry to bother you, we went back to using the polyphase resample since the mono-phase takes a very long time to convert even quite small files which does admittedly make a lot of sense. I have implemented the changes you suggested (see first response), however it still does not seem to accept 32khz files and was wondering if you have been able to find anything wrong with the library, though I would be more than willing to suspect it was a failure on my end causing the issue.

Thanks 

Ewan Stacey BTO

**Reply by psambit9791** [reply]
Hey,

Sorry I have not had a chance to look into this yet. I'll get to it as soon as I can.

**Comment by psambit9791** [comment, marked as answer]
Hi @Ewan-Stacey-BTO-IS ,

I have used the file that you had shared and **Polyphase Resampling** is working as expected. 

Sharing the properties here:

```
Source Signal Properties: 
Sample Rate: 32000
Channels: 1
Number of Points: 128000
```

```
Desired Signal Properties: 
Sample Rate: 22050
Channels: 1
Number of Points: 88200
```

Which if you calculate: **(22050/32000) * 128000 is 88200**

Also attaching the data (input and output) as a ZIP file here.
[data.zip](https://github.com/psambit9791/jdsp/files/14320688/data.zip)


Finally, the sample code I used is here:
```java
package org.example;

import com.github.psambit9791.jdsp.io.WAV;
import com.github.psambit9791.jdsp.misc.UtilMethods;
import com.github.psambit9791.jdsp.signal.Resample;
import com.github.psambit9791.wavfile.WavFileException;
import org.apache.commons.math3.util.ArithmeticUtils;

import java.io.IOException;
import java.util.Hashtable;

public class Main {
    public static void printProps(Hashtable h1, double[] signal, String signalType) {
        System.out.println(signalType + " Properties: ");
        System.out.println("Sample Rate: " + h1.get("SampleRate"));
        System.out.println("Channels: " + h1.get("Channels"));
        System.out.println("Number of Points: " + signal.length);
        System.out.println("\n\n");
    }

    public static void main(String[] args) throws IOException, WavFileException {

        // Reading the Source Audio File and getting the properties
        WAV objReadSource = new WAV();
        objReadSource.readWAV("data/audio_32000.wav");
        double[][] signal = objReadSource.getData("int");
        double[] signalFlat = UtilMethods.flattenMatrix(signal);
        Hashtable<String, Long> propsOut = objReadSource.getProperties();
        printProps(propsOut, signalFlat, "Source Signal");

        // This is where the key operation happens
        long desiredSampleRate = 22050;
        long sourceSampleRate = propsOut.get("SampleRate");
        long maxFrequency = ArithmeticUtils.lcm(sourceSampleRate, desiredSampleRate);
        int upScaler = (int)(maxFrequency/sourceSampleRate);
        int downScaler = (int)(maxFrequency/desiredSampleRate);

        Resample pr = new Resample(upScaler, downScaler, "constant");
        double[] newSignalFlat = pr.resampleSignal(signalFlat);
        double[][] newSignal = new double[newSignalFlat.length][1];
        for (int i=0; i<newSignalFlat.length; i++) {
            newSignal[i][0] = newSignalFlat[i];
        }

        // Writing the resampled signal into a new WAV file
        WAV objWrite = new WAV();
        String outputFilename = "data/audio_22050.wav";
        objWrite.putData(newSignal, (long)desiredSampleRate, "int", outputFilename);

        // Reading the new WAV file and getting its properties for comparison
        WAV objReadDesired = new WAV();
        objReadDesired.readWAV("data/audio_22050.wav");
        double[][] signalDesired = objReadDesired.getData("int");
        double[] signalDesiredFlat = UtilMethods.flattenMatrix(signalDesired);
        Hashtable<String, Long> propsOutDesired = objReadDesired.getProperties();
        printProps(propsOutDesired, signalDesiredFlat, "Desired Signal");
    }
}
```

I have also run the same code on a WAV file with a duration of *02:48* (which is where the key issue had been encountered according to your first post) and that has worked fine as well. Sharing the input and output files for your reference as a zip attachment.
[two_minutes_plus_test.zip](https://github.com/psambit9791/jdsp/files/14320713/two_minutes_plus_test.zip)


Can you please check this out and let me know if you are still encountering the issue? 
<ins>And if this works, please highlight what had caused the issue for posterity and please mark this as the answer.</ins>

**Reply by Ewan-Stacey-BTO-IS** [reply]
Hey, after some testing this does seem to work though you will need to split up larger files since this only seems to work for smaller files.

1. From what I can tell the issue was being caused by the number of samples being entered into the resampler, causing issues specifically when asked to resample 8800000 samples at 32khz I don't have any idea why since it could do 10000000 samples from 32khz to 22.05khz and I haven't been able to test if any other sample sizes fail
2. I implemented a new system using your solution which goes by the sample rate instead which looks like this
```Java

long desiredSampleRateLong = desiredSampleRate.longValue();
long sourceSampleRate = (long) originalFormat.getSampleRate();
long maxFrequency = ArithmeticUtils.lcm(sourceSampleRate, desiredSampleRateLong);
int upScaler = (int)(maxFrequency/sourceSampleRate);
int downScaler = (int)(maxFrequency/desiredSampleRateLong);
                
Resample resampler = new Resample(upScaler, downScaler, "constant");
double[] preResamplingArray = convertFloatArrayToDouble(trimmedValues);
trimmedValues = new float[0];
int maxArrayResampleSize = (int)sourceSampleRate * 4;
ArrayList<double[]> splitResampledArrays = new ArrayList<>();
int outerArraySize = (int) Math.ceil((double) preResamplingArray.length / (double) maxArrayResampleSize);

for (int x = 0; x < outerArraySize; x++) {
    int endOfRange;
                    
     if (x == (outerArraySize - 1)) {
         endOfRange = preResamplingArray.length;
      } else {
          endOfRange = ((x + 1) * maxArrayResampleSize);
      }

    double[] splitArray = Arrays.copyOfRange(preResamplingArray, x * maxArrayResampleSize, endOfRange);
    double[] resampledDoubleArray = resampler.resampleSignal(splitArray);
    splitResampledArrays.add(resampledDoubleArray);
}
                
double[] recombinedDoubleArray = recombineDoubleArray(splitResampledArrays);
resampledValues = convertDoubleArrayToFloat(recombinedDoubleArray);
```
